src/App.tsx

```diff
@@ -1,24 +1,214 @@
+    }
+  };
+
+  useEffect(() => {
+    fetchTodosFromServer();
+  }, [tempTodo, loadingTodoIds, visibleError]);
+
+  const completedTodoIds = useMemo(() => (
+    getcompletedTodoIds(todos)
+  ), [todos]);
+
+  const visibleTodos = useMemo(() => (
+    preparedTodos(todos, todoFilter)
+  ), [todos, todoFilter]);
+
+  const removeTodos = useCallback(async (todoIds: number[]) => {
+    try {
+      setLoadingTodoIds(prevIds => [...prevIds, ...todoIds]);
+
+      await Promise.all(
+        todoIds.map(async id => {
+          await removeTodo(id);
+        }),
+      );
+
+      setTodos((previousTodos) => (
+        previousTodos.filter(todo => !todoIds.includes(todo.id))
+      ));
+    } catch (error) {
+      setVisibleError('Unable to delete a todo');
+    } finally {
+      setLoadingTodoIds([]);
+    }
+  }, []);
+
+  const handleUpdate = useCallback(async (
+    todoIds: number[],
+    newTitle?: string,
+  ) => {
+    try {
+      setLoadingTodoIds((prevIds) => [...prevIds, ...todoIds]);
+
+      await Promise.all(
+        todoIds.map(async id => {
+          const todoToUpdate = todos.find(todo => todo.id === id);
+
+          if (todoToUpdate) {
+            const updatedTodo = await changeTodo(id, {
+              ...todoToUpdate,
+              completed: newTitle
+                ? todoToUpdate.completed
+                : !todoToUpdate.completed,
+              title: newTitle || todoToUpdate.title,
+            });
+
+            setTodos(prevTodos => prevTodos.map(todo => (
+              todo.id === updatedTodo.id
+                ? updatedTodo
+                : todo
+            )));
+          }
+        }),
+      );
+    } catch (error) {
+      setVisibleError('Unable to update a todo');
+    } finally {
+      setLoadingTodoIds([]);
+    }
+  }, [todos]);
+
+  const addTodo = useCallback(async (title: string) => {
+    try {
+      const newTodo = {
+        title: title.trim(),
+        completed: false,
+        userId: USER_ID,
+      };
+
+      const tempTodoId = 0;
+
+      setTempTodo({
+        ...newTodo,
+        id: tempTodoId,
+      });
+
+      setLoadingTodoIds([tempTodoId]);
+
+      const addedTodo = await addTodoToServer('/todos', newTodo);
+
+      setTodos((currentTodos) => [...currentTodos, addedTodo]);
+    } catch (error) {
+      setVisibleError('Unable to add a todo');
+    } finally {
+      setTempTodo(null);
+      setLoadingTodoIds([]);
+      setTodoTitle('');
+    }
+  }, []);
 
-export const App: React.FC = () => {
   if (!USER_ID) {
     return <UserWarning />;
   }
 
+  const handleSubmit = (
+    event: FormEvent<HTMLFormElement>,
+  ) => {
+    event.preventDefault();
+
+    if (!todoTitle.trim()) {
+      setTodoTitle('');
+      setVisibleError('Title can\'t be empty');
+
+      return;
+    }
+
+    addTodo(todoTitle);
+  };
+
+  const handleQueryChange = (
+    event: ChangeEvent<HTMLInputElement>,
+  ) => {
+    setTodoTitle(event.target.value);
+  };
```

Comment:
Would be better to create these 2 handlers inside the component that has an input and a form.
Provide only the `addTodo()` function inside your `header` component